src/main/java/com/hierynomus/smbj/paths/SymlinkPathResolver.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (C)2016 - SMBJ Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hierynomus.smbj.paths;
+
+import com.hierynomus.mserref.NtStatus;
+import com.hierynomus.mssmb2.SMB2Error;
+import com.hierynomus.mssmb2.SMB2Functions;
+import com.hierynomus.mssmb2.SMB2Packet;
+import com.hierynomus.smbj.common.SmbPath;
+import com.hierynomus.smbj.session.Session;
+
+import java.nio.charset.StandardCharsets;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Set;
+
+import static com.hierynomus.utils.Strings.join;
+import static com.hierynomus.utils.Strings.split;
+
+public class SymlinkPathResolver implements PathResolver {
+    private PathResolver wrapped;
+    private Set<NtStatus> states;
+
+    public SymlinkPathResolver(PathResolver wrapped) {
+        this.wrapped = wrapped;
+        this.states = EnumSet.copyOf(wrapped.handledStates());
+        this.states.add(NtStatus.STATUS_STOPPED_ON_SYMLINK);
+    }
+
+    @Override
+    public SmbPath resolve(Session session, SMB2Packet responsePacket, SmbPath smbPath) throws PathResolveException {
+        if (responsePacket.getHeader().getStatus() == NtStatus.STATUS_STOPPED_ON_SYMLINK) {
+            SMB2Error.SymbolicLinkError symlinkData = getSymlinkErrorData(responsePacket.getError());
+            if (symlinkData == null) {
+                throw new PathResolveException(responsePacket.getHeader().getStatus(), "Create failed for " + smbPath + ": missing symlink data");
+            }
+            String target = resolveSymlinkTarget(smbPath.getPath(), symlinkData);
+            return new SmbPath(smbPath.getHostname(), smbPath.getShareName(), target);
+        }
+
+        return wrapped.resolve(session, responsePacket, smbPath);
+    }
+
+    @Override
+    public Set<NtStatus> handledStates() {
+        return states;
```

Comment:
This should probably be a defensive copy or be wrapped in an unmodifiable collection. Otherwise the internal state of the PathResolver is mutable.